feat: storybook-builder and storybook-framework-web-components

bashmish opened the pull request:
Fixes #2025

## Questions

- [ ] mark as beta state? in docs or/and in package version
- [ ] what to do with old plugin docs and guide? mark as outdate and link to Migration guide?
- [ ] make a guide now? do we need a guide?

## TODOs

- [ ] fix `npm run update:mjs-dts-entrypoints`

Westbrook commented:
I love it! Especially all the docs you've already included 👨🏼‍🎤 

Things I know we need before something like this could land:
- I know there are naming conventions in the SB ecosystem, but shoudl this be @web/dev-server-storybook@2.0?
- a working version of @web/dev-server-storybook@1.x, 😞, regardless of which name we use. Skipping 1.x all together _could_ be an option, but I'm not willing to make that decision for myself, just yet. Hopefully @koddsson can make some progress on that, soon 🤞🏼 
- if we don't use @web/dev-server-storybook
  - WDS Storybook needs to be deprecated _officially_. That means no code/docs/etc. remain.
  - should the new packages jump to 1.0? 
- CI needs to pass
  - right now it's red, but likely we need more things, too
  - we need to run Storybook in CI
  - we need to build Storybook in CI
  - should it just be kitchen sink storybook, or do we need to do various add-on combinations?
  - should Storybook do anything else in CI?
- is it possible to not ship a CJS configuration? 

Other than the code _mostly_ working for me, I've not looked to closely at it, but once we start getting answers to the above, I am excited to do so!

LarsDenBakker commented on packages/storybook-builder/src/generate-stories-script.ts:
```diff
@@ -0,0 +1,57 @@
+      logger.warn(`Cannot process ${ext} file with storyStoreV7: ${relativePath}`);
+    }
+
+    return `  '${toImportPath(relativePath)}': () => import('${process.cwd()}/${relativePath}')`;
```
Might need to take care of windows file paths in dynamic imports here?

bashmish replied:
thanks, I'll normalize the right path too

bashmish replied:
I'm not sure it's actually needed, I checked what happens then in `packages/dev-server-rollup/src/rollupAdapter.ts` and seems like the path will be normalized before doing the check which resolves this path correctly, but there are a few checks that are done on the non-normalized one before it which might go wrong

I also doubt now that the left path is normalized good for Windows

I wish I could debug on Windows myself and see if there is a problem and where, but on M1 I can't, so until that I'd rather keep things as is, or I'll have to guess both the problem and solution :D

the long-term plan is to have automated tests and run a Storybook in the build, so I think we can also run one in our Windows pipeline

shilman commented:
This looks amazing -- incredible job @bashmish !!! Is there an easy way for me to test this out?

Westbrook commented:
@shilman we're testing the dev release in SWC right now (via the releases made against @bashmish's local packages)
- [branch](https://github.com/adobe/spectrum-web-components/tree/storybook)
- [preview](https://storybook--spectrum-web-components.netlify.app/storybook/?path=/story/accordion-densities-compact--s)

Only one [weirdness](https://github.com/adobe/spectrum-web-components/blob/storybook/patches/%40storybook%2Bblocks%2B7.4.0.patch) in Storybook to make it work.

bashmish commented:
> I love it! Especially all the docs you've already included 👨🏼‍🎤

@Westbrook Thanks for the compliment!

> Things I know we need before something like this could land:
> 
> * I know there are naming conventions in the SB ecosystem, but shoudl this be @web/dev-server-storybook@2.0?

`@web/dev-server-storybook` is a plugin for WDS which connects Storybook to it
`@web/storybook-builder` is a plugin for Storybook which connects WDS to it
so the relationship between WDS and Storybook has turned around with the Builders architecture

our own convention for WDS plugins is `@web/dev-server-plugin-name`, but `@web/storybook-builder` is not a plugin for WDS, so I thought it shouldn't be `@web/dev-server-storybook` or start with `@web/dev-server-`

For the same reason I made a new top section in the documentation, instead of fitting it into the WDS as a subsection.

Storybook package naming convention were hard to follow too after moving inside modern-web:
I could just add `@web` prefix and make
`@web/storybook-builder-wds`
`@web/storybook-web-components-wds`
but
- in other packages we don't add `wds`, e.g. instead of `@web/test-runner-wds` we just use `@web/test-runner`
  - so I renamed `@web/storybook-builder-wds` to `@web/storybook-builder`
  - and first renamed `@web/storybook-web-components-wds` to `@web/storybook-web-components`
- in other plugins we usually put the grouping name in the front, e.g. `@web/dev-server-plugin-name`
  - so I decided to have a clear grouping name for frameworks (like `rollup-pugin-`) so that they are all ordered next to each other and renamed further `@web/storybook-web-components` to `@web/storybook-framework-web-components`

> * a working version of @web/[dev-server-storybook@1.x](mailto:dev-server-storybook@1.x), 😞, regardless of which name we use. Skipping 1.x all together _could_ be an option, but I'm not willing to make that decision for myself, just yet. Hopefully @koddsson can make some progress on that, soon 🤞🏼

I think @koddsson is busy fixing it

> * if we don't use @web/dev-server-storybook
>   
>   * WDS Storybook needs to be deprecated _officially_. That means no code/docs/etc. remain.

This is also one of my questions in the PR description
I agree we should deprecate it, but not sure we should remove docs right away. Maybe keep it in place with links to Migration for the time being?

>   * should the new packages jump to 1.0?

I think too early, because some of the features are yet to be implemented, mainly:
- CLI for init
- CLI for migration
- MDX
- MDJS
- HMR

> * CI needs to pass
>   
>   * right now it's red, but likely we need more things, too

Yeah, some TS errors which I didn't have locally. I'll make sure to fix them.

>   * we need to run Storybook in CI
>   * we need to build Storybook in CI
>   * should it just be kitchen sink storybook, or do we need to do various add-on combinations?
>   * should Storybook do anything else in CI?

Fully agree on that. Recent issues with 1.x show the urgency for it.
We can learn from Storybook monorepo how to test it, I saw they have generators for various combinations, which is a must for them to test multiple builders and frameworks. Though we have only one now, so can simplify it for the time being.

> * is it possible to not ship a CJS configuration?

This week I tested the ESM version and couldn't make it work, but I didn't investigate why yet.
The only thing I saw is that Storybook expected a CJS module of a framework.
IIRC, Storybook has plans to make all dependencies loaded in the browser to be ESM in Storybook 8 which is a lot of work, but on the Node level I don't see why we can't make it work already now, especially given `.storybook/` config files support ESM now.
This is not the highest prio for me right now.

> 
> Other than the code _mostly_ working for me, I've not looked to closely at it, but once we start getting answers to the above, I am excited to do so!

Great! I'm open to discuss anything to make sure this is reviewable and goes in the right direction.

shilman commented:
FWIW weirdness fixed here https://github.com/storybookjs/storybook/releases/tag/v7.5.0-alpha.0

bashmish commented:
@Westbrook you can also remove https://github.com/adobe/spectrum-web-components/blob/storybook/patches/%40chialab%2Bestransform%2B0.17.4.patch and update to `@chialab/estransform@0.17.5` as it was fixed in there:
- chialab/rna#157
- chialab/rna#156

thepassle commented on docs/docs/storybook-builder/configuration.md:
```diff
@@ -0,0 +1,131 @@
+# Storybook Builder >> Configuration ||2
+
+Web Storybook Builder aims at compatibility with standard Storybook features, e.g. configuration of preview (`.storybook/preview.js`), official addons, etc...
```
```suggestion
`@web/storybook-builder` aims at compatibility with standard Storybook features, e.g. configuration of preview (`.storybook/preview.js`), official addons, etc.
```

bashmish replied:
I tried to follow the style of the other doc pages and use a product name rather than package name. Do you think it should be the package?

bashmish replied:
for the same reason also used `Web Dev Server` instead of `@web/dev-server`

thepassle replied:
Aha ok. Hm, I do think the package name is more clear. I don't think its clear to everyone that `Web Dev Server` -> `@web/dev-server`. A "Web Dev Server" is pretty generic

thepassle replied:
ok I'll leave the decision up to you then, I do think `@web/dev-server` is more clear though

bashmish replied:
I'm OK to update then, since I had this concern myself about unclear names

would be nice to also make sure we use these names consistently in other parts of the docs

thepassle commented on docs/docs/storybook-builder/configuration.md:
```diff
@@ -0,0 +1,131 @@
+
+## Configuring builder and framework
+
+Set type and framework in `main.js`:
```
```suggestion
Set the `type` and `framework` in `main.js`:
```

thepassle commented on docs/docs/storybook-builder/configuration.md:
```diff
@@ -0,0 +1,131 @@
+
+## Configuring dev server
+
+The builder implements a backend for `storybook dev` CLI command and comes with a default Web Dev Server configuration which adheres to browser standards and supports official addons of the Storybook ecosystem.
```
```suggestion
The builder implements a backend for the `storybook dev` CLI command and comes with a default `@web/dev-server` configuration which adheres to browser standards and supports official addons of the Storybook ecosystem.
```

thepassle commented on docs/docs/storybook-builder/configuration.md:
```diff
@@ -0,0 +1,131 @@
+
+### Web Dev Server configuration file
+
+When Storybook is loaded, the builder's default Web Dev Server config gets automatically merged with the project's [configuration file (`web-dev-server.config.{mjs,cjs,js}`)](../dev-server/cli-and-configuration.md#configuration-file) if present.
```
```suggestion
When Storybook is loaded, the builder's default `@web/dev-server` config gets automatically merged with the project's [configuration file (`web-dev-server.config.{mjs,cjs,js}`)](../dev-server/cli-and-configuration.md#configuration-file) if present.
```

thepassle commented on docs/docs/storybook-builder/configuration.md:
```diff
@@ -0,0 +1,131 @@
+
+### Option `wdsFinal`
+
+Sometimes you might need to add Storybook specific configuration for dev server, you can use the `wdsFinal` option for this:
```
```suggestion
Sometimes you might need to add Storybook specific configuration for the dev server, you can use the `wdsFinal` option for this:
```

thepassle commented on docs/docs/storybook-builder/configuration.md:
````diff
@@ -0,0 +1,131 @@
+export default config;
+```
+
+The `config` parameter is a Web Dev Server config, make sure to use appropriate options and Web Dev Server plugins.
````
```suggestion
The `config` parameter is a `@web/dev-server` config option, you can use this to customize your `@web/dev-server` options and plugins.
```

thepassle commented on docs/docs/storybook-builder/configuration.md:
````diff
@@ -0,0 +1,131 @@
+```
+
+The `config` parameter is a Web Dev Server config, make sure to use appropriate options and Web Dev Server plugins.
+When using rollup plugins make sure to [convert them to Web Dev Server ones](../dev-server/plugins/rollup.md).
````
```suggestion
When using rollup plugins make sure to [convert them to `@web/dev-server` ones](../dev-server/plugins/rollup.md).
```

thepassle commented on docs/docs/storybook-builder/configuration.md:
````diff
@@ -0,0 +1,131 @@
+export default config;
+```
+
+The `config` parameter is a `rollup` config, make sure to use appropriate options and rollup plugins.
````
```suggestion
The `config` parameter is a `rollup` config, you can use this to customize your `rollup` options and plugins.
```

thepassle commented on docs/docs/storybook-builder/frameworks.md:
```diff
@@ -0,0 +1,36 @@
+# Storybook Builder >> Frameworks ||3
+
+Storybook supports different technologies via [frameworks](https://storybook.js.org/docs/web-components/api/new-frameworks).
+It simplifies the configuration of the Storybook for a specific builder like Web Storybook Builder.
```
```suggestion
It simplifies the configuration of the Storybook for a specific builder like `@web/storybook-builder`.
```

thepassle commented on docs/docs/storybook-builder/migration-to-storybook-7.md:
```diff
@@ -0,0 +1,117 @@
+# Storybook Builder >> Migration to Storybook 7 ||4
+
+This guide explains how to migrate from [`@web/dev-server-storybook` plugin](../dev-server/plugins/storybook.md) (that used opinionated Storybook 6 bundle `@web/storybook-prebuilt`) to Storybook 7 and new Web Storybook Builder.
```
```suggestion
This guide explains how to migrate from [`@web/dev-server-storybook` plugin](../dev-server/plugins/storybook.md) (that used opinionated Storybook 6 bundle `@web/storybook-prebuilt`) to Storybook 7 and new `@web/storybook-builder`.
```

thepassle commented on docs/docs/storybook-builder/migration-to-storybook-7.md:
```diff
@@ -0,0 +1,117 @@
+The Storybook CLI `upgrade` command is not aware of `@web/dev-server-storybook` plugin and as a result it won't work by default.
+But with a simple workaround you can make it work.
+
+First install old version of `builder-vite` and old Storybook 6 renderer for the technology you used in the old setup, e.g. for Web Components it will be:
```
```suggestion
First, install the old version of `builder-vite` and old Storybook 6 renderer for the technology you used in the old setup, e.g. for Web Components it will be:
```

thepassle commented on docs/docs/storybook-builder/migration-to-storybook-7.md:
````diff
@@ -0,0 +1,117 @@
+npx storybook@latest upgrade
+```
+
+Then [configure builder and framework](./configuration.md#configuring-builder-and-framework) in the main Storybook configuration.
````
```suggestion
Then [configure the builder and framework](./configuration.md#configuring-builder-and-framework) in the main Storybook configuration.
```

thepassle commented on docs/docs/storybook-builder/migration-to-storybook-7.md:
```diff
@@ -0,0 +1,117 @@
+
+## Rename rollupConfig to rollupFinal
+
+For consistency with other similar hooks in the Storybook ecosystem, including Web Storybook Builder `wdsFinal`, the rollup hook was renamed to `rollupFinal`.
```
```suggestion
For consistency with other similar hooks in the Storybook ecosystem, including `@web/storybook-builder`'s `wdsFinal`, the rollup hook was renamed to `rollupFinal`.
```

thepassle commented on docs/docs/storybook-builder/migration-to-storybook-7.md:
```diff
@@ -0,0 +1,117 @@
+
+## Usage of ESM
+
+Storybook 7 supports ES modules for configuration files, so we recommend using standard syntax for imports and exports in `.storybook/main.js` and other configuration files.
```
```suggestion
Storybook 7 supports ES modules for configuration files, so we recommend using standard ESM syntax for imports and exports in `.storybook/main.js` and other configuration files.
```

thepassle commented on docs/docs/storybook-builder/overview.md:
```diff
@@ -0,0 +1,33 @@
+
+# Web Storybook Builder
+
+Web Storybook Builder is a [Storybook builder](https://storybook.js.org/docs/web-components/builders/overview) powered by [Web Dev Server](../dev-server/overview.md). If you are using the ecosystem of Web Dev Server and Web Test Runner, then this is the right choice for your Storybook setup.
```
```suggestion
`@web/storybook-builder` is a [Storybook builder](https://storybook.js.org/docs/web-components/builders/overview) powered by [Web Dev Server](../dev-server/overview.md). If you are using the ecosystem of Web Dev Server and Web Test Runner, then this is the right choice for your Storybook setup.
```

thepassle commented on docs/docs/storybook-builder/overview.md:
```diff
@@ -0,0 +1,33 @@
+
+## Installation
+
+Install the Web Storybook Builder:
```
```suggestion
Install the `@web/storybook-builder`:
```

thepassle commented on packages/storybook-builder/src/rollup-plugin-prebundle-modules.ts:
```diff
@@ -0,0 +1,190 @@
+  'unfetch',
+  'util-deprecate',
+  'vue',
+  'warning',
```
This is the list that was supposed to have shrunk already right? Or is this a different list?

bashmish replied:
it's the old one
new one I have in a local branch only and needs more testing
I'd prefer to work on it after finishing this

thepassle replied:
fine by me 👍  just checking

koddsson commented:
> > * a working version of @web/[dev-server-storybook@1.x](mailto:dev-server-storybook@1.x), 😞, regardless of which name we use. Skipping 1.x all together _could_ be an option, but I'm not willing to make that decision for myself, just yet. Hopefully @koddsson can make some progress on that, soon 🤞🏼
> 
> I think @koddsson is busy fixing it
> 
> > *

It _should_ all be fixed now :godmode: 🤞🏻

thepassle commented:
lgtm :)
